test/model/tools/model_test/main.cpp

```diff
@@ -283,75 +284,60 @@ load_workload(const char *file)
 }
 
 /*
- * reduce_counterexample --
- *     Try to find a smaller workload that reproduces a failure.
+ * reduce_counterexample_context_t --
+ *     The context for counterexample reduction.
  */
-static void
-reduce_counterexample(std::shared_ptr<model::kv_workload> workload, const std::string &main_home,
-  const std::string &reduce_home, const std::string &conn_config, const std::string &table_config)
-{
+struct reduce_counterexample_context_t {
+    const std::string &conn_config;
+    const std::string &main_home;
+    const std::string &reduce_home;
+    const std::string &table_config;
+
+    size_t round;
+
     /*
-     * Separate the workload back into sequences, where a sequence is a transaction or just a single
-     * non-transactional operations, such as database restart or set stable timestamp. We will not
-     * use the sequences directly; our main goal is to annotate each workload operation with the
-     * corresponding sequence number, which we will then use to filter the list of operations.
-     *
-     * Note that unlike the sequences generated by the workload generator, the list of sequences
-     * will not form an equivalent serial schedule, because here we order transactions by their
-     * begin operations, not by commits.
+     * reduce_counterexample_context_t::reduce_counterexample_context_t --
+     *     Initialize the context.
      */
-    std::unordered_map<model::txn_id_t, model::kv_workload_sequence_ptr> txn_to_sequence;
-    std::deque<model::kv_workload_sequence_ptr> sequences;
+    reduce_counterexample_context_t(const std::string &main_home, const std::string &reduce_home,
+      const std::string &conn_config, const std::string &table_config)
+        : main_home(main_home), reduce_home(reduce_home), conn_config(conn_config),
+          table_config(table_config), round(0)
+    {
+    }
+};
 
+/*
+ * reduce_counterexample_by_aspect --
+ *     Try to find a smaller workload that reproduces a failure, focusing on a specific aspect
+ *     specified by the corresponding lambda arguments for detecting and extracting the operation
+ *     aspect (e.g., its sequence number or its table ID) and any other condition under which the
+ *     operation should be always included in the tested workload. The parameters to the lambdas are
+ *     the operation itself and its index in the workload. Return the reduced workload, or the
+ *     workload parameter itself if no reduction has been found.
+ */
+template <typename T>
+static std::shared_ptr<model::kv_workload>
+reduce_counterexample_by_aspect(reduce_counterexample_context_t &context,
+  std::shared_ptr<model::kv_workload> workload, const char *aspect_name_plural,
```

Comment:
Can `aspect_name_plural` be a `const std::string&`?

Reply:
It would work, but the intention is for the caller to provide a string literal, which is of type `const char*`. So I think that may be a better fit?